### PR TITLE
Remove `full` from `similar(full(X), T, dims)` calls in generic concatenation methods

### PR DESCRIPTION
**Pull request by Sacha0**
Towards reintroducing JuliaLang/julia#17079, this pull request removes `full` from `similar(full(X), T, dims)` calls in the generic concatenation methods.

For an explanation of those `full` calls' purpose, see the first paragraph of https://github.com/JuliaLang/julia/pull/17079#issuecomment-234658313. See the preceding comments in that thread to understand why displacing those `full` calls is necessary to reintroduce JuliaLang/julia#17079 .

Contrary to the expectation I expressed in that comment, removing those `full` calls appears to _improve_ the degree to which the generic concatenation methods preserve structure:

**<details>**The code in [this gist](https://gist.github.com/Sacha0/e7fdbb3dc61840e8df9c339cdc28f0a1#file-typemapconcat-jl) determines, for all pairwise combinations of dense vectors, dense matrices, sparse vectors, sparse matrices, special matrices, and triangular, symmetric, and hermitian annotations of both dense and sparse matrices, the return type of prototypical `vcat`, `hcat`, `hvcat`, and `cat` calls, and prints the mappings `catmeth(typeA, typeB) -> typeRet`. [This gist](https://gist.github.com/Sacha0/e7fdbb3dc61840e8df9c339cdc28f0a1#file-master-concattypemap-jl) provides the output of that code on master, and [this gist](https://gist.github.com/Sacha0/e7fdbb3dc61840e8df9c339cdc28f0a1#file-defull-concattypemap-jl) the same with this PR. [This gist](https://gist.github.com/Sacha0/e7fdbb3dc61840e8df9c339cdc28f0a1#file-diff-concattypemap-diff) provides the diff. (Update: These gists are up to date, so the diff is now empty.)**</details>**

**tl;dr of details** To sum up the diff: When a sparse matrix or vector appears as the first argument in a concatenation call for which there exists no specialized method, on master the result is an `Array`, whereas with this PR the result remains sparse. Otherwise the results are identical.

This change makes JuliaLang/julia#2326 more apparent in that, for example, `hcat(SparseMatrixCSC, Bidiagonal)` would now produce a sparse matrix whereas `hcat(Bidiagonal, SparseMatrixCSC)` would still produce a dense matrix. Modifying the `Union`s in [this and the accompanying method definitions](https://github.com/JuliaLang/julia/blob/e717ded47a55d3526a10207d8d75ae5c9ab9646f/base/sparse/sparsematrix.jl#L3234) to include special matrix types would resolve that issue for special matrix types. But the same cannot be done for the annotation types. And absent the ability to write methods with `Vararg`s in positions other than last, I do not immediately see a way to resolve that issue for annotation types. (Edit: Ref. JuliaLang/julia#13130. Combinatorial explosion precludes extending the present strategy much?) Making the change necessary to handle special matrix types while punting to JuliaLang/julia#2326 on annotation types seems attractive. Other solutions? Thanks!

### CODE REVIEW

**Comment by tkelman** [comment]
x-ref the related https://github.com/JuliaLang/julia/pull/15172 and https://github.com/JuliaLang/julia/pull/16722

**Comment by martinholters** [comment]
I'd also like to point at #16740 (once I get to working on that again), see especially https://github.com/JuliaLang/julia/pull/16740#issuecomment-230715239, where `hcat` (and friends) basically just decide the desired return type and then forward to `typed_hcat` (or similar). Hopefully that simplifies writing specializations to pick the correct return type, and if that still fails, allows the caller to override it as needed.

**Comment by Sacha0** [comment]
AV i686 failure looks like #17662?

**Comment by Sacha0** [comment]
I've updated the gists following #17675 and #17680. Only one set of discrepancies remain: When the first argument is a `SparseMatrixCSC` or `SparseVector` and the second argument is an annotation type (`<:AbstractTriangular`, `Symmetric`, or `Hermitian`), on master the result is dense whereas with this PR the result is sparse. #17684 causes this.

**Comment by Sacha0** [comment]
Gists updated following #17684. The diff is now empty, so this PR and subsequent reinstatement of #17079 should be viable now. (I will cherry pick #17079 back together if / when this clears, unless there is a better approach to reinstating #17079?). Best!

**Comment by tkelman** [comment]
let's keep it busy @nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17660#issuecomment-247210781) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/b4d5ca7b41bbebf3ca54bb45b3bece121fdf1a24/666f91e_vs_fdf3748/report.md). cc @jrevels

**Comment by Sacha0** [comment]
Nanosoldier seems pleased. All well? Thanks!

**Comment by Sacha0** [comment]
Absent objections or requests for time, I plan to merge this Monday (PDT). Best!

**Comment by tkelman** [comment]
Is this going to break DataArrays again or has it been refactored to avoid relying on this?

**Comment by Sacha0** [comment]
Removing these `full(X)` calls should not break DataArrays; removal (rather than replacement with `convert(Array, X)`) was the (hypothetical) fix for https://github.com/JuliaStats/DataArrays.jl/issues/208. But testing DataArrays on this branch prior to merge would be wise. @martinholters, do you have min repro code for https://github.com/JuliaStats/DataArrays.jl/issues/208 that I could test? Thanks!

**Comment by martinholters** [comment]
Too bad `Pkg.test("DataArrays")` fails on current master. I guess
```Julia
using DataArrays
T = Float64
n = 1000
a = Array(T, n)
na = bitrand(n)
nna = sum(na)
da = DataArray(a, na)
isequal(sort(da), [DataArray(sort(dropna(da))); DataArray(T, nna)])
```
should be the relevant part of the test that failed last time.

**Comment by Sacha0** [comment]
Thanks @martinholters! Results on this branch are identical to those on master. Good to merge, @tkelman @martinholters? Best!

**Comment by Sacha0** [comment]
Absent objections or requests for time, I plan to merge this Friday (PDT). Best!

**Comment by martinholters** [comment]
No objections from my side.
